File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        res = []
        prod = 1
        pnz = 1
        z = 0
        for num in nums:
            if num != 0:
                pnz *= num
            else:
                z += 1
            prod *= num
        
        for num in nums:
            if num == 0:
                if z > 1:
                    res.append(0)
                else:
                    res.append(pnz)
            else:
                res.append(int(prod/num))
        return res

File: Data Structures & Algorithms/sliding-window-maximum/submission-0.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {

        vector<int> res;
        priority_queue<int> maxHeap;
        unordered_map<int, int> idx;

        for(int i = 0; i < k; i++){
            maxHeap.push(nums[i]);
            idx[nums[i]] = i;
        }

        res.push_back(maxHeap.top());

        for(int r = k; r < nums.size(); r++){
            int l = r-k+1;
            idx[nums[r]] = r;
            maxHeap.push(nums[r]);

            while(idx[maxHeap.top()] < l){
                maxHeap.pop();
            }

            res.push_back(maxHeap.top());
        }

        return res;
    }
};
